### PR TITLE
Reimplement WebRTC controls as a hook

### LINKED ISSUE #594: Switching between desktop and mobile layout causes you to leave chat
This is of course because the CallSection component gets unmounted/remounted without carrying state across, but we should see if we can do better.

### PR DESCRIPTION
**Pull request by zarvox**
This replaces the hierarchy of components that managed our WebRTC
interactions with a new `useCallState` hook which encapsulates all of
the call state management functionality.  This provides a key benefit:
we can hoist the actual state storage higher in the component tree,
which means that we won't lose call state when switching layouts between
desktop and mobile screen widths.

As a bonus, this change also corrects a bunch of things from the audio
calls implementation that would have gone awry under `StrictMode` in
React 18 due to doing side-effectful actions outside of `useEffect`
blocks.

Fixes #594.

### CODE REVIEW

File: imports/client/components/PuzzlePage.tsx
Comment (zarvox) [inline imports/client/components/PuzzlePage.tsx:71]: whoops, I should switch that back to `false`

File: imports/client/components/ChatPeople.tsx
Comment (zarvox) [inline imports/client/components/ChatPeople.tsx:275]: This is taken care of

Comment (zarvox) [reply imports/client/components/ChatPeople.tsx]: Specifically, by https://github.com/deathandmayhem/jolly-roger/pull/1023/commits/fb9d6c23d7f6a3a1c2870f7cad8703ad2f7a267a#diff-abb257e3f8aad9487024c922792f027ffc5ec9de52e60cadb72c6e439d6ef4a3R324-R333

File: imports/client/hooks/useCallState.ts
Comment (zarvox) [inline imports/client/hooks/useCallState.ts:335]: The reason we handle this subscription (and the `mediasoup:transports` sub below) manually instead of using `useTracker` or `useSubscription` is because of an unfortunate interaction between:

* us using these publishes to cause server-side state changes
* us expecting `[huntId, puzzleId, tabId]` (or below, `peer._id`) to generally be unique per client, with last-writer-wins semantics on conflict
* React reserving the right to rerun `render()` multiple times, as exercised by `StrictMode`
* `useTracker` (and as a result `useSubscription`) running the inner function body during both the render phase and the effect phase, resulting in multiple subscriptions

The solution is to move our use of state-changing subscriptions to run exclusively in the effect phase, which is reasonably straightforward.

In hindsight, this probably deserves a comment.

File: imports/client/hooks/useCallState.ts
Comment (zarvox) [inline imports/client/hooks/useCallState.ts:448]: clean up

File: imports/client/hooks/useCallState.ts
Comment (zarvox) [inline imports/client/hooks/useCallState.ts:571]: It should be possible to phrase this without `!`

File: imports/client/hooks/useCallState.ts
Comment (zarvox) [inline imports/client/hooks/useCallState.ts:653]: actually do this

File: imports/client/components/AudioConfig.tsx
Comment (zarvox) [inline imports/client/components/AudioConfig.tsx:265]: I slipped this (and the matching change in `CallSection`) into this PR since I saw it's what's causing #1029
to fail presubmit checks at https://github.com/deathandmayhem/jolly-roger/runs/8193203773?check_suite_focus=true and this change rewrites the line in `CallSection` anyway.

The backstory is kinda fun -- this is a vestige of the distant past from when I was bringing up the mesh WebRTC implementation with video because I didn't yet have anything like a spectrogram and I was sick of hearing feedback when the connection was successfully established.  Once I had the implementation working reliably, I switched the `video` element to an `audio` one, without realizing this property isn't valid for `audio`.

File: imports/client/hooks/useCallState.ts
Comment (ebroder) [inline imports/client/hooks/useCallState.ts:352]: This will fire on the new tracks, right? Do you want to capture a reference to the old `mediaSource` in the `useEffect` before the unwind callback?

File: imports/client/hooks/useCallState.ts
Comment (ebroder) [inline imports/client/hooks/useCallState.ts:523]: I think I understand it after reading the rest of the file, but I think I'd appreciate a comment here about what these generation numbers trigger

**Review by ebroder — APPROVED** [review]
Sorry for not getting to this sooner. I'm a little sad that we can't rely on React to manage more of the lifecycle for us, but given that they explicitly seem to not want to do that, this seems like it will be a better long-term structure.

I left a few comments. Feel free to address them (or not, if I misunderstood something). I'll approve now so you don't need to wait for another review from me.

File: imports/client/hooks/useCallState.ts
Comment (zarvox) [reply imports/client/hooks/useCallState.ts:343]: I think it's a good idea, but the obvious implementation doesn't appear to do anything because we need to hook into react-router instead, since it's not a browser-level navigation that any of our links trigger.

This was something that was included in react-router v5, but is absent in v6: https://reactrouter.com/en/main/upgrading/v5#prompt-is-not-currently-supported .  Given that this looks like it'll be a bit more work to sort out, I filed #1037 for future followup.

File: imports/client/hooks/useCallState.ts
Comment (zarvox) [reply imports/client/hooks/useCallState.ts:352]: Ah, that's probably wise, thanks.  In practice, I think the way we've written the rest of our code means `mediaSource` never transitions from truthy to truthy, so we never hit this edge case, but it's good to be resilient.

File: imports/client/hooks/useCallState.ts
Comment (zarvox) [reply imports/client/hooks/useCallState.ts:523]: Wrote up a bit more about how the producer flow works, including what fields these generation counters are a proxy for changing, and which effects read those fields (and thus, need to rerun when the generation counter changes).

**Pull request merged**
